modules/mesh-layers/src/utils/matrix.js

```diff
@@ -0,0 +1,125 @@
+};
+
+const rotationMatrix = new Float32Array(16);
+const modelMatrix = new Matrix4();
+
+function updateRotationMatrix({roll, pitch, yaw}) {
+  const sr = Math.sin(roll);
+  const sp = Math.sin(pitch);
+  const sw = Math.sin(yaw);
+
+  const cr = Math.cos(roll);
+  const cp = Math.cos(pitch);
+  const cw = Math.cos(yaw);
+
+  rotationMatrix[0] = cw * cp; // 0,0
+  rotationMatrix[1] = sw * cp; // 1,0
+  rotationMatrix[2] = -sp; // 2,0
+  rotationMatrix[3] = 0;
+  rotationMatrix[4] = -sw * cr + cw * sp * sr; // 0,1
+  rotationMatrix[5] = cw * cr + sw * sp * sr; // 1,1
+  rotationMatrix[6] = cp * sr; // 2,1
+  rotationMatrix[7] = 0;
+  rotationMatrix[8] = sw * sr + cw * sp * cr; // 0,2
+  rotationMatrix[9] = -cw * sr + sw * sp * cr; // 1,2
+  rotationMatrix[10] = cp * cr; // 2,2
+  rotationMatrix[11] = 0;
+  rotationMatrix[12] = 0;
+  rotationMatrix[13] = 0;
+  rotationMatrix[14] = 0;
+  rotationMatrix[15] = 1;
+}
+
+function makeMatrix({object, getYaw, getPitch, getRoll, getScale, getTranslation}) {
+  const roll = getRoll(object) * RADIAN_PER_DEGREE;
+  const pitch = getPitch(object) * RADIAN_PER_DEGREE;
+  const yaw = getYaw(object) * RADIAN_PER_DEGREE;
+  const scale = getScale(object);
+  const translate = getTranslation(object);
+
+  updateRotationMatrix({roll, pitch, yaw});
+
+  modelMatrix
+    .identity()
+    .translate(translate)
+    .multiplyRight(rotationMatrix)
+    .scale(scale);
+
+  return modelMatrix;
+}
+
+function calculateInstanceModelMatrix(attribute) {
+  // eslint-disable-next-line no-invalid-this
+  const {data, getYaw, getPitch, getRoll, getScale, getTranslation, getMatrix} = this.props;
+
+  const instanceModelMatrixData = attribute.value;
+
+  let i = 0;
+  for (const object of data) {
+    const matrix =
+      getMatrix(object) ||
+      makeMatrix({object, getYaw, getPitch, getRoll, getScale, getTranslation});
+
+    instanceModelMatrixData[i++] = matrix[0];
+    instanceModelMatrixData[i++] = matrix[1];
+    instanceModelMatrixData[i++] = matrix[2];
+    instanceModelMatrixData[i++] = matrix[3];
+    instanceModelMatrixData[i++] = matrix[4];
+    instanceModelMatrixData[i++] = matrix[5];
+    instanceModelMatrixData[i++] = matrix[6];
+    instanceModelMatrixData[i++] = matrix[7];
+    instanceModelMatrixData[i++] = matrix[8];
+    instanceModelMatrixData[i++] = matrix[9];
+    instanceModelMatrixData[i++] = matrix[10];
+    instanceModelMatrixData[i++] = matrix[11];
+    instanceModelMatrixData[i++] = matrix[12];
+    instanceModelMatrixData[i++] = matrix[13];
+    instanceModelMatrixData[i++] = matrix[14];
+    instanceModelMatrixData[i++] = matrix[15];
+  }
+}
+
+export function addMatrixToAttributeManager(attributeManager) {
+  attributeManager.addInstanced({
+    instanceModelMatrix: {
+      size: 16,
+      accessor: ['getYaw', 'getPitch', 'getRoll', 'getScale', 'getTranslation', 'getMatrix'],
+      shaderAttributes: {
+        instanceModelMatrix__LOCATION_0: {
```

Comment:
I always try to avoid pushing too much logic into utils. I want the top-level code to remain declarative and easy to read.

Therefore maybe just export a constant here and keep the addInstanced call in the layers.

```js
const MATRIX_SHADER_ATTRIBUTES = {
        instanceModelMatrix__LOCATION_1: {
          size: 4,
          stride: 64,
          offset: 0
        },
        instanceModelMatrix__LOCATION_1: {
          size: 4,
          stride: 64,
          offset: 16
        },
        instanceModelMatrix__LOCATION_2: {
          size: 4,
          stride: 64,
          offset: 32
        },
        instanceModelMatrix__LOCATION_3: {
          size: 4,
          stride: 64,
          offset: 48
        }
```